blog/_posts/2015-09-27-auto-diff-in-julia.md

```diff
@@ -0,0 +1,257 @@
+
+julia> deriv1 = deriv(d)
+1.589002649374538
+
+julia> using Calculus; deriv2 = Calculus.derivative(f, 1.0)
+1.5890026493377403
+
+julia> deriv1 - deriv2
+3.679767601738604e-11
+```
+
+Notice that our dual number result comes *close* to the result obtained from Calculus.jl, but is actually slightly different. That slight difference is due to the approximation error inherent to the finite differencing method employed by Calculus.jl.
+
+In reality, the number types that ForwardDiff.jl implements are quite a bit more complicated than `DualNumber`. `ForwardDiffNumber`s behave like **ensembles** of dual numbers and [hyper-dual numbers (the higher-order analog of dual numbers)](https://adl.stanford.edu/hyperdual/Fike_AIAA-2011-886.pdf), allowing for simultaneous calculation of multiple higher-order partial derivatives. For an in-depth examination of ForwardDiff.jl's number type implementation, see [this section of the developer documentation](http://www.juliadiff.org/ForwardDiff.jl/types.html).
+
+## Performance Comparison: The Ackley Function
+
+To illustrate the performance gains that can be achieved using ForwardDiff.jl, let's do a (very) naive benchmark comparison between ForwardDiff.jl and other libraries. For this comparison, we'll be comparing the time to calculate the gradient of a function using ForwardDiff.jl, Calculus.jl, and a Python-based AD tool, AlgoPy.
+
+### Defining the Ackley Function
+
+The function we'll be using in our test is the [Ackley function](http://www.sfu.ca/~ssurjano/ackley.html), which is mathematically defined as
+
+$$f(\vec{x}) = -a \exp( -b \sqrt{\frac{1}{k} \sum_{i=1}^k x^{2}_{i}}) - \exp(\frac{1}{k} \sum_{i=1}^k \cos(cx_{i})) + a + \exp(1)$$
+
+Here's the definition of the function in Julia:
+
+```julia
+function ackley(x)
+    a, b, c = 20.0, -0.2, 2.0*π
+    len_recip = inv(length(x))
+    sum_sqrs = zero(eltype(x))
+    sum_cos = sum_sqrs
+    for i in x
+        sum_cos += cos(c*i)
+        sum_sqrs += i^2
+    end
+    return (-a * exp(b * sqrt(len_recip*sum_sqrs)) -
+            exp(len_recip*sum_cos) + a + e)
+end
+```
+
+...and here's the corresponding Python definition:
+
+```python
+def ackley(x):
+    a, b, c = 20.0, -0.2, 2.0*numpy.pi
+    len_recip = 1.0/len(x)
+    sum_sqrs, sum_cos = 0.0, 0.0
+    for i in x:
+        sum_cos += algopy.cos(c*i)
+        sum_sqrs += i*i
+    return (-a * algopy.exp(b*algopy.sqrt(len_recip*sum_sqrs)) -
+            algopy.exp(len_recip*sum_cos) + a + numpy.e)
+```
+
+### Benchmark Results
+
+The benchmarks were performed with input vectors of length 16, 1600, and 16000, taking the best time out of 5 trials for each test.
+
+#### Function evaluation time
+
+The below table compares the evaluation times of `ackley(x)` in both Python and Julia:
+
+| length(x) | Python time (s) | Julia time (s) | Speed-Up vs. Python  |
+|-----------|-----------------|----------------|----------------------|
+| 16        | 0.00011         | 2.294e-6       | 47.95x               |
```

Comment:
That's not a bad idea. If I was going to redo these benchmarks, I'd probably also want to use [Benchmarks.jl](https://github.com/johnmyleswhite/Benchmarks.jl) instead of `@time`. Then I'd have to redo the other benchmarks as well. 

Honestly, I was thinking of redoing these anyway since I stupidly forgot to document the Julia/Python versions I was using for the initial run...

Unfortunately, I'm not sure I have time to do all this before tomorrow, which is apparently when the JSoC posts are "due" (@ViralBShah is that a hard deadline?).

> (optionally taking the mean, afterwards)

The reason I used "best" instead of average was to ensure that oddities like occasional GC didn't skew the results.